# Tasks
### This workbook contains my solutions to the tasks for the Machine Learning and Statistics Module





#### Task 1 - October 5th, 2020 
Write a Python function called sqrt2 that calculates and prints to the screen the square root of 2 to 100 decimal places. Your code should not depend on any module from the standard library or otherwise. You should research the task first and include references and a description of your algorithm.
#### Research
Newton's method [1] can be used to iteratively approach closer toware the square root of a number. According to Newton's method, one can approach further toward the square root of a number, x, from an initial guess of the square root, z, as follows:

$$ z_{next} = z - \frac{z^2-x}{2z} $$

There are two impediments to achieving a precision to 100 decimal places without using any Python modules, however:
1. Decimal point numbers do not have exact representations in binary floating point, i.e. in Python's float datatype, which is the only datatype available for displaying decimal numbers in Python without the use of modules such as Decimal [2]. This means that we will not likely be able to achieve the accuracy required to display the square root of two to 100 decimal places, as in order to do so we would have to perform many very precise calculations involving floating point numbers, which are by definition imprecise.
2. Calculating to 100 decimal places is computationally expensive and is not feasible on standard laptops with limited RAM and CPU cores.

What will thus been shown as part of my solution is as follows:
1. Python's Math module's sqrt function is not accurate to the 100th decimal place (we will use NASA calculation of the root of 2 to ~ million decimal places as our comparison [3]), nor is it able to provide a 100 decimal places. From this we can infer that the inaccuracy resides in teh datatype used by this function, namely, the float.
2. Python's Decimal module's sqrt function is accurate to 100 decimal places, again indicating that Python's float datatype is what is at issue above.
3. The computational expense of calculating a square root using only in-built functions (i.e. no imported functions) increases significantly when the difference between the actual root squared and the calculated root squared is less than .0001.

The conclusion from this is that to calculate the square root of 2 in Python one must import a module, e.g. the Decimal module.

[1] https://mathworld.wolfram.com/NewtonsMethod.html

[2] https://docs.python.org/2/library/decimal.html

[3] https://stackoverflow.com/questions/22162522/how-to-display-a-decimal-number-to-100-decimal-places

In [1]:
# hint at the inaccuracy of the sqrt function in the
# Math module, which uses floats
from math import sqrt
sqrt(2)**2

2.0000000000000004

In [2]:
# print the Math module's square root of 2 to 100
# decimal places
answer = "%.100f" % sqrt(2)
print(answer)
# confirm that there are 100 decimal places
print(len(str((answer))) - 2)

1.4142135623730951454746218587388284504413604736328125000000000000000000000000000000000000000000000000
100


In [3]:
# we will take NASA's calculation of the root of 2
# as accurate to 100 decimal places
nasa = "1.41421356237309504880168872420969807856967187537694807317667973799073247846210703885038753432764157273501384623091229702492483605585073721264412149709993583141322266592750559275579995050115278206057147010955997160597027453459686201472851..."
nasa = nasa[:101]
# check at which decimal place the Math module
# becomes inaccurate, and thus floats likely
# become inaccurate
for i in range(102):
    if str(answer)[i] != nasa[i]:
        accuracy = i - 2
        break
# check how many places (inaccurate or otherwise)
# the math module can calculate the root of 2 to
for i in range(102):
    if str(answer)[i] != "0":
        x = 0
    if str(answer)[i] == "0":
               x+=1
    if x == 4:
        place = i-2-x
        break
print("Python's Math module's sqrt function is accurate to the " + str(accuracy) + "th place, although it provides an answer to the " + str(place) + "th place.")

Python's Math module's sqrt function is accurate to the 15th place, although it provides an answer to the 51th place.


In [4]:
# show that the Decimal module can calculate the
# square root of 2 to 100 decimal places, by 
# comparing with NASA's figure, above
from decimal import Decimal, getcontext
# set the decimal point precision to 100
getcontext().prec = 100
# get the decimal square root of 3
Decimal(2).sqrt()
print(nasa)
print(Decimal(2).sqrt())
# as the NASA has not been rounded, 
# we will check the first 99 decimal values
# for simplicity instead
print(str(Decimal(2).sqrt())[:100] == nasa[:100])

1.414213562373095048801688724209698078569671875376948073176679737990732478462107038850387534327641572
1.414213562373095048801688724209698078569671875376948073176679737990732478462107038850387534327641573
True


In [5]:
# Find the difference between 2 and the square root of 2
# to a hundred decimal places squared
# note that this number is far smaller than that for
# the Math module's function (.0000000000000004)
Decimal(2 - 1.414213562373095048801688724209698078569671875376948073176679737990732478462107038850387534327641573 ** 2)

Decimal('-4.44089209850062616169452667236328125E-16')

In [6]:
# design an algorithm implementing Newton's Method
# Newton's method: better guess = 0.5 * (guess + number / guess)
# store the desired precision in a variable
# we decide on the precision outside of the function itself
# as the requirements of the question do not indicate
# that precision should be inputted as a parameter to the function
precision = 0.01
def squareRootOfTwo():
    # have the function take a guess
    guess = 1.5
    # while the guess is not of a sufficient accuracy,
    # apply Newton's algorithm to improve it.
    # To check the accuracy of the guess, square it and
    # check if the absolute difference between the result 
    # and the input is less than the desired precision.
    while abs(guess ** 2 - 2) > precision:
        # if greater than the precision, apply Newton's alogrithm
        guess = guess - ((guess ** 2 - 2) / 2 * guess)
        # once the guess is sufficiently accurate the while
        # loop ends. Now round the result to your liking
    print("%.100f" % guess)
    return str(answer)

In [7]:
# Now we will test Newton's method for calculating
# the square of two, and time the calculation
import time
start = time.time()
answer = squareRootOfTwo()
end = time.time()
print(str(round(end - start, 1)) + " seconds taken")
# check how many places (inaccurate or otherwise)
# this run of the algorithm calculated the root of 2 to
for i in range(len(answer)):
    if str(answer)[i] != "0":
        x = 0
    if str(answer)[i] == "0":
               x+=1
    if x == 4:
        place = i-2-x
        break
print("Running our function so that the square of the calculated square root of two is between 1.9 and 2.1 provides an approximation of the square root of two to the " + str(place) + "th place.")

1.4177445877362315762582056777318939566612243652343750000000000000000000000000000000000000000000000000
0.0 seconds taken
Running our function so that the square of the calculated square root of two is between 1.9 and 2.1 provides an approximation of the square root of two to the 51th place.


In [8]:
# Same again but to a precision of 0.00001
# note the increase from 89.6 seconds to seconds
import time
precision = 0.001
start = time.time()
answer = squareRootOfTwo()
end = time.time()
print(str(round(end - start, 1)) + " seconds taken")
# check how many places (inaccurate or otherwise)
# this run of the algorithm calculated the root of 2 to
for i in range(len(answer)):
    if str(answer)[i] != "0":
        x = 0
    if str(answer)[i] == "0":
               x+=1
    if x == 4:
        place = i-2-x
        break
print("Running our function so that the square of the calculated square root of two is between 1.99 and 2.01 provides an approximation of the square root of two to the " + str(place) + "th place.")

1.4145670714723390659628421417437493801116943359375000000000000000000000000000000000000000000000000000
1.1 seconds taken
Running our function so that the square of the calculated square root of two is between 1.99 and 2.01 provides an approximation of the square root of two to the 51th place.


In [9]:
# Same again but to a precision of 0.00001
# note the increase from 89.6 seconds to seconds
import time
precision = 0.0001
start = time.time()
answer = squareRootOfTwo()
end = time.time()
print(str(round(end - start, 1)) + " seconds taken")
# check how many places (inaccurate or otherwise)
# this run of the algorithm calculated the root of 2 to
for i in range(len(answer)):
    if str(answer)[i] != "0":
        x = 0
    if str(answer)[i] == "0":
               x+=1
    if x == 4:
        place = i-2-x
        break
print("Running our function so that the square of the calculated square root of two is between 1.999 and 2.001 provides an approximation of the square root of two to the " + str(place) + "th place.")

1.4142489172699921340381479240022599697113037109375000000000000000000000000000000000000000000000000000
104.6 seconds taken
Running our function so that the square of the calculated square root of two is between 1.999 and 2.001 provides an approximation of the square root of two to the 51th place.


We can note two things from the running of our function with different precision demands:
1. The number of decimal places remains the same (51) despite varying precision demands, indicating that there is a limit to the number of decimal places that a float is capable of storing.
2. The computational expense of calculating the square root of two increases exponentially with respect to an increased demand for precision (from 0.0 seconds at 0.1 precision to 1.1 seconds at 0.001 precision to approximately 100 seconds at 0.0001 precision).

*The conclusion from this is thus that it is neither computationally feasibly for a typical laptop, nor possible due to the nature of Python's float datatype, to calculate the square root of 2 to 100 decimal places without importing any modules.*